kmod/core/core.c

```diff
@@ -583,29 +586,64 @@ static int kpatch_kallsyms_callback(void *data, const char *name,
 {
 	struct kpatch_kallsyms_args *args = data;
 
-	if (args->addr == addr && !strcmp(args->name, name))
+	if ((mod && !args->objname) || (!mod && args->objname))
+		return 0;
+
+	if (strcmp(args->name, name))
+		return 0;
+
+	if (args->objname && strcmp(args->objname, mod->name))
+		return 0;
+
+	args->addr = addr;
+	args->count++;
+
+	/*
+	 * Finish the search when the symbol is found for the desired position
+	 * or the position is not defined for a non-unique symbol.
+	 */
+	if ((args->pos && (args->count == args->pos)) ||
+	    (!args->pos && (args->count > 1))) {
 		return 1;
+	}
 
 	return 0;
 }
 
-static int kpatch_verify_symbol_match(const char *name, unsigned long addr)
+static int kpatch_find_object_symbol(const char *objname, const char *name,
+				     unsigned long sympos, unsigned long *addr)
 {
-	int ret;
-
 	struct kpatch_kallsyms_args args = {
+		.objname = objname,
 		.name = name,
-		.addr = addr,
+		.addr = 0,
+		.count = 0,
+		.pos = sympos,
 	};
 
-	ret = kallsyms_on_each_symbol(kpatch_kallsyms_callback, &args);
-	if (!ret) {
-		pr_err("base kernel mismatch for symbol '%s'\n", name);
-		pr_err("expected address was 0x%016lx\n", addr);
-		return -EINVAL;
+	mutex_lock(&module_mutex);
+	kallsyms_on_each_symbol(kpatch_kallsyms_callback, &args);
+	mutex_unlock(&module_mutex);
+
+	/*
+	 * Ensure an address was found. If sympos is 0, ensure symbol is unique;
+	 * otherwise ensure the symbol position count matches sympos.
+	 */
+	if (args.addr == 0)
+		pr_err("symbol '%s' not found in symbol table\n", name);
+	else if (args.count > 1 && sympos == 0) {
+		pr_err("unresolvable ambiguity for symbol '%s' in object '%s'\n",
+		       name, objname);
+	} else if (sympos != args.count && sympos > 0) {
+		pr_err("symbol position %lu for symbol '%s' in object '%s' not found\n",
+		       sympos, name, objname ? objname : "vmlinux");
+	} else {
+		*addr = args.addr;
+		return 0;
 	}
 
-	return 0;
+	*addr = 0;
+	return -EINVAL;
 }
 
 static unsigned long kpatch_find_module_symbol(struct module *mod,
```

Comment:
@flaming-toast : in this case, isn't ```kpatch_find_external_symbol``` going to pick them up as an exported symbol by way of ``find_symbol```?  I need to test the "it's in another .o within the patch module" case.